File: leetcode/solutions/0383.RansomNote/canConstruct.go
package solution

func canConstruct(ransomNote string, magazine string) bool {
	if len(ransomNote) > len(magazine) {
		return false
	}

	letterCount := make([]int, 26)
	base := int32('a')

	for _, r := range magazine {
		letterCount[r-base]++
	}

	for _, r := range ransomNote {
		letterCount[r-base]--

		if letterCount[r-base] < 0 {
			return false
		}
	}

	return true
}

File: leetcode/solutions/0383.RansomNote/canConstruct.ts
export default function canConstruct(
  ransomNote: string,
  magazine: string
): boolean {
  const charCount: number[] = Array.from({ length: 26 }, () => 0);
  const base = "a".charCodeAt(0);

  for (let i = 0; i < magazine.length; i++) {
    charCount[magazine.charCodeAt(i) - base]++;
  }

  for (let i = 0; i < ransomNote.length; i++) {
    const index = ransomNote.charCodeAt(i) - base;
    charCount[index]--;

    if (charCount[index] < 0) {
      return false;
    }
  }

  return true;
}
